File: src/main/java/dev/cele/asa_sm/ui/components/forms/NumberField.java
package dev.cele.asa_sm.ui.components.forms;

import com.formdev.flatlaf.FlatClientProperties;
import lombok.Getter;

import javax.swing.*;
import java.util.ArrayList;
import java.util.Objects;
import java.util.function.Consumer;

public class NumberField extends JTextField {

    private final ArrayList<Consumer<Number>> numberListeners = new ArrayList<>();

    @Getter
    private Number number = 0;

    @Getter
    private boolean isInteger = true;
    public void setIsInteger(boolean isInteger){
        this.isInteger = isInteger;
        if(isInteger){
            setNumber(number.intValue());
        }
    }

    public void setNumber(Number number){
        if(!Objects.equals(number, this.number)){
            numberListeners.forEach(listener -> listener.accept(number));
            this.number = number;
            updateText();
        }
    }

    private void updateText(){
        setText("" + (isInteger ? number.intValue() : number));
    }

    public NumberField() {
        this(0);
    }

    public NumberField(Number value) {
        super(5);

        //add focus lost listener
        addFocusListener(new java.awt.event.FocusAdapter() {
            public void focusLost(java.awt.event.FocusEvent evt) {
                validateTextAndUpdateNumber();
            }
        });

        //add listener for ENTER key
        addActionListener(e -> validateTextAndUpdateNumber());
    }

    public void validateTextAndUpdateNumber(){
        var text = getText();
        try {
            number = isInteger ? Integer.parseInt(text) : Double.parseDouble(text);
            numberListeners.forEach(listener -> listener.accept(number));
        } catch (NumberFormatException e) {
            //TODO: make a beep?
            putClientProperty(FlatClientProperties.OUTLINE, "warning");
            if(threadCancelWarning != null){
                threadCancelWarning.interrupt();
            }
            threadCancelWarning = new Thread(() -> {
                try {
                    Thread.sleep(1000);
                } catch (InterruptedException interruptedException) {
                    return;
                }
                putClientProperty(FlatClientProperties.OUTLINE, null);
            });
            threadCancelWarning.start();
            setText(number.toString());
        }
    }

    Thread threadCancelWarning = null;

    public void addNumberListener(Consumer<Number> listener){
        numberListeners.add(listener);
    }
    public void removeNumberListener(Consumer<Number> listener){
        numberListeners.remove(listener);
    }

}

File: src/main/java/dev/cele/asa_sm/ui/components/NumberField.kt
package dev.cele.asa_sm.ui.components

import dev.cele.asa_sm.ui.addValueChangedListener
import javax.swing.JTextField
import javax.swing.SwingUtilities

class NumberField(startingValue: Int): JTextField() {

    var value: Int = startingValue
        get() = field
        set(value) {
            field = value
            SwingUtilities.invokeLater {
                text = value.toString()
            }
        }

    var valueChangeListeners: MutableList<(Int) -> Unit> = mutableListOf()
    fun addValueChangedListener(function: (Int) -> Unit) {
        valueChangeListeners.add(function)
    }

    fun removeValueChangedListener(function: (Int) -> Unit) {
        valueChangeListeners.remove(function)
    }

    init {
        this.value = startingValue

        addKeyListener(object : java.awt.event.KeyAdapter() {
            override fun keyReleased(e: java.awt.event.KeyEvent?) {
                super.keyReleased(e)
                try{
                    value = Integer.parseInt(text)
                }catch (e: NumberFormatException){
                    //if the text is not a number set back the old value and return
                    text = value.toString()
                    return
                }
                valueChangeListeners.forEach { it(value) }
            }
        })
    }

}
